pandas/core/indexes/base.py

```diff
@@ -7156,6 +7155,43 @@ def shape(self) -> Shape:
         return (len(self),)
 
 
+def maybe_sequence_to_range(sequence) -> Any | range:
+    """
+    Convert a 1D sequence to a range if possible.
+
+    Returns the input if not possible.
+
+    Parameters
+    ----------
+    sequence : 1D sequence
+    names : sequence of str
+
+    Returns
+    -------
+    Any : input or range
+    """
+    if hasattr(sequence, "dtype") and not isinstance(sequence, np.ndarray):
+        return sequence
+    np_sequence = np.asarray(sequence)
+    if np_sequence.dtype.kind != "i" or len(sequence) == 1:
+        return sequence
+    elif len(sequence) == 0:
+        return range(0)
+    diff = np_sequence[1] - np_sequence[0]
```

Comment:
Why are only the first two elements important here?

Reply:
We're taking an example diff here (just between the first 2 elements) and later seeing if the rest of the diffs between the rest of the elements match